### PR TITLE
SNOW-1161484: Use insertRows instead of insertRow for schematization

### PR DESCRIPTION
**Pull request by sfc-gh-tzhang**
# Overview

SNOW-1161484

- Create channel with SKIP_BATCH instead of CONTINUE, to avoid the case that KC can crash right after adding the good rows to table, and the bad rows will be missing in the DLQ
- Update the schematization code to use insertRows instead of insertRow
        - Better performance
        - If KC is configured with a longer buffer flush time, everything will be ingested as one batch so that the 1 second flush or 32MB channel size won't apply

## Pre-review checklist
- [ ] This change should be part of a Behavior Change Release. See [go/behavior-change](http://go/behavior-change).
- [ ] This change has passed Merge gate tests
- [ ] Snowpipe Changes
- [ ] Snowpipe Streaming Changes
- [ ] This change is TEST-ONLY
- [ ] This change is README/Javadocs only
- [ ] This change is protected by a config parameter <PARAMETER_NAME> eg `snowflake.ingestion.method`.
    - [ ] `Yes` - Added end to end and Unit Tests. 
    - [ ] `No` - Suggest why it is not param protected
- [ ] Is his change protected by parameter <PARAMETER_NAME> on the server side?
    - [ ] The parameter/feature is not yet active in production (partial rollout or PrPr, see [Changes for Unreleased Features and Fixes](http://go/ppp-prpr)).
    - [ ] If there is an issue, it can be safely mitigated by turning the parameter off. This is also verified by a test (See [go/ppp](http://go/ppp)).

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 1]
```diff
@@ -1059,7 +1059,7 @@ private SnowflakeStreamingIngestChannel openChannelForTable() {
             .setDBName(this.sfConnectorConfig.get(Utils.SF_DATABASE))
             .setSchemaName(this.sfConnectorConfig.get(Utils.SF_SCHEMA))
             .setTableName(this.tableName)
-            .setOnErrorOption(OpenChannelRequest.OnErrorOption.CONTINUE)
+            .setOnErrorOption(OpenChannelRequest.OnErrorOption.SKIP_BATCH)
```
Comment (sfc-gh-xhuang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1062]: This will be a BCR?

Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1062]: I don't think so, there is no behavior difference as far as customer is concerned

**Review by sfc-gh-xhuang — APPROVED** [review]

**Comment by sfc-gh-japatel** [comment]
Could you consider using the checklist in PR? Let me know if you want to change something. 

Also is SKIP_BATCH needed for schematization or is needed in general? If needed in general, I would like if it be separated from this PR.

**Review by sfc-gh-tjones — APPROVED** [review]

**Comment by sfc-gh-xhuang** [comment]
I would merge this into the next release if it is ready but up to you

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 2]
```diff
@@ -576,6 +584,22 @@ InsertRowsResponse insertBufferedRecords(StreamingBuffer streamingBufferToInsert
...
+    for (long rowIdx = 0; rowIdx < streamingBufferToInsert.getNumOfRecords(); rowIdx++) {
+      if (errorIdx < insertErrors.size() && rowIdx == insertErrors.get(errorIdx).getRowIndex()) {
+        errorIdx++;
+      } else {
+        buffer.insert(streamingBufferToInsert.getSinkRecord(rowIdx));
+      }
+    }
+    return buffer;
+  }
+
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:602]: will this result into a validation error you added recently?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 3]
```diff
@@ -576,6 +584,22 @@ InsertRowsResponse insertBufferedRecords(StreamingBuffer streamingBufferToInsert
     return response;
   }
 
+  /** Building a new buffer which contains only the good rows from the original buffer */
+  private StreamingBuffer rebuildBufferWithoutErrorRows(
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:588]: is there a test coverage for this new method?

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java [hunk 4]
```diff
@@ -686,6 +686,23 @@ public void testInsertRowsWithSchemaEvolution() throws Exception {
 
       topicPartitionChannel.insertBufferedRecordsIfFlushTimeThresholdReached();
 
+      // Verify that the buffer is cleaned up and nothing is in DLQ because of schematization error
+      Assert.assertTrue(topicPartitionChannel.isPartitionBufferEmpty());
+      Assert.assertEquals(0, kafkaRecordErrorReporter.getReportedRecords().size());
+
+      // Do it again without any schematization error, and we should have row in DLQ
```
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:693]: i am confused about the comment, do you mean not have a row in DLQ if there is no schematization error?

or do you mean to verify the DLQ results something along this lines? Assert.assertEquals(0, kafkaRecordErrorReporter.getReportedRecords().size());?

**Review by sfc-gh-japatel — COMMENTED** [review]
I am worried about this change not having enough tests, do you feel existing tests are enough? 

Also, highly encourage for atleast important changes to fill out PR checklist. Thanks

**Comment by sfc-gh-japatel** [comment]
> Could you consider using the checklist in PR? Let me know if you want to change something in the template.
> 
> Also is SKIP_BATCH needed for schematization or is needed in general? If needed in general, I would like if it be separated from this PR.

talked offline, dont need to split PRs, SKIP_BATCH is needed and there is no behavior change.

**Review by sfc-gh-japatel — COMMENTED** [review]
left couple of comments, I dont have any more concerns, will approve in next iteration! Thank you, this will lower schematization latency 🥇

**Comment by sfc-gh-xhuang** [comment]
@sfc-gh-rcheng @sfc-gh-tzhang this should be merged for 2.2.2 release

**Review by sfc-gh-rcheng — APPROVED** [review]
I'm also a little concerned about this big of a behavior change without additional tests, but looks like @sfc-gh-japatel discussed this with toby offline.

**Comment by sfc-gh-xhuang** [comment]
With the latest, ingest-sdk merged. We should be able to update the ingest-sdk version and merge this PR?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 5]
```diff
@@ -657,52 +681,28 @@ public InsertRowsResponse get() throws Throwable {
...
-                offsets.get(idx) - this.insertRowsStreamingBuffer.getFirstOffset();
-            if (extraColNames == null && nonNullableColumns == null) {
-              InsertValidationResponse.InsertError newInsertError =
-                  new InsertValidationResponse.InsertError(
-                      insertError.getRowContent(), originalSinkRecordIdx);
-              newInsertError.setException(insertError.getException());
-              newInsertError.setExtraColNames(insertError.getExtraColNames());
-              newInsertError.setMissingNotNullColNames(insertError.getMissingNotNullColNames());
-              // Simply added to the final response if it's not schema related errors
-              finalResponse.addError(insertError);
```
Comment (sfc-gh-wtrefon) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:690]: @sfc-gh-tzhang  why you removed this part? IMO we still need it to properly add the rows to the rebuilt buffer and also send only the non-schema errors into DLQ

**Comment by sfc-gh-wtrefon** [comment]
Closed in favor of https://github.com/snowflakedb/snowflake-kafka-connector/pull/866 due to merge conflicts

**Pull request closed without merging**